Source/hydro/trace_plm.cpp

```diff
@@ -0,0 +1,297 @@
+      qm(i,j+1,k,QUN) = un_ref + (apleft - amleft)*cc/rho;
+      qm(i,j+1,k,QUT) = ut_ref + azut1left;
+      qm(i,j+1,k,QUTT) = utt_ref + azutt1left;
+      qm(i,j+1,k,QPRES) = max(lsmall_pres, p_ref + (apleft + amleft)*csq);
+      qm(i,j+1,k,QREINT) = rhoe_ref + (apleft + amleft)*enth*csq + azeleft;
+
+      // add the source terms
+      qm(i,j+1,k,QRHO) = amrex::max(lsmall_dens, qm(i,j+1,k,QRHO) + 0.5_rt*dt*srcQ(i,j,k,QRHO));
+      qm(i,j+1,k,QUN) += 0.5_rt*dt*srcQ(i,j,k,QUN);
+      qm(i,j+1,k,QUT) += 0.5_rt*dt*srcQ(i,j,k,QUT);
+      qm(i,j+1,k,QUTT) += 0.5_rt*dt*srcQ(i,j,k,QUTT);
+      qm(i,j+1,k,QREINT) += 0.5_rt*dt*srcQ(i,j,k,QREINT);
+      qm(i,j+1,k,QPRES) += 0.5_rt*dt*srcQ(i,j,k,QPRES);
+
+    } else if (idir == 2 && k <= vhi[2]) {
+      qm(i,j,k+1,QRHO) = amrex::max(lsmall_dens, rho_ref + apleft + amleft + azrleft);
+      qm(i,j,k+1,QUN) = un_ref + (apleft - amleft)*cc/rho;
+      qm(i,j,k+1,QUT) = ut_ref + azut1left;
+      qm(i,j,k+1,QUTT) = utt_ref + azutt1left;
+      qm(i,j,k+1,QPRES) = amrex::max(lsmall_pres, p_ref + (apleft + amleft)*csq);
+      qm(i,j,k+1,QREINT) = rhoe_ref + (apleft + amleft)*enth*csq + azeleft;
+
+      // add the source terms
+      qm(i,j,k+1,QRHO) = amrex::max(lsmall_dens, qm(i,j,k+1,QRHO) + 0.5_rt*dt*srcQ(i,j,k,QRHO));
+      qm(i,j,k+1,QUN) += 0.5_rt*dt*srcQ(i,j,k,QUN);
+      qm(i,j,k+1,QUT) += 0.5_rt*dt*srcQ(i,j,k,QUT);
+      qm(i,j,k+1,QUTT) += 0.5_rt*dt*srcQ(i,j,k,QUTT);
+      qm(i,j,k+1,QREINT) += 0.5_rt*dt*srcQ(i,j,k,QREINT);
+      qm(i,j,k+1,QPRES) += 0.5_rt*dt*srcQ(i,j,k,QPRES);
+
+    }
+
+#if (AMREX_SPACEDIM < 3)
+    // geometry source terms -- these only apply to the x-states
+    if (idir == 0 && dloga(i,j,k) != 0.0_rt) {
+      Real courn = dtdx*(cc + abs(un));
+      Real eta = (1.0_rt-courn)/(cc*dt*abs(dloga(i,j,k)));
+      Real dlogatmp = amrex::min(eta, 1.0_rt)*dloga(i,j,k);
+      Real sourcr = -0.5_rt*dt*rho*dlogatmp*un;
+      Real sourcp = sourcr*csq;
+      Real source = sourcp*enth;
+
+      if (i <= vhi[0]) {
+        qm(i+1,j,k,QRHO) += sourcr;
+        qm(i+1,j,k,QRHO) = amrex::max(qm(i+1,j,k,QRHO), lsmall_dens);
+        qm(i+1,j,k,QPRES) += sourcp;
+        qm(i+1,j,k,QREINT) += source;
+      }
+
+      if (i >= vlo[0]) {
+        qp(i,j,k,QRHO) += sourcr;
+        qp(i,j,k,QRHO) = amrex::max(qp(i,j,k,QRHO), lsmall_dens);
+        qp(i,j,k,QPRES) += sourcp;
+        qp(i,j,k,QREINT) += source;
+      }
+    }
+#endif
+
+    for (int ipassive = 0; ipassive < npassive; ipassive++) {
+      int n = qpass_map_p[ipassive];
+
+      // Right state
+      if ((idir == 0 && i >= vlo[0]) ||
+          (idir == 1 && j >= vlo[1]) ||
+          (idir == 2 && k >= vlo[2])) {
+
+        Real spzero = un >= 0.0_rt ? -1.0_rt : un*dtdx;
+        qp(i,j,k,n) = q_arr(i,j,k,n) + 0.5_rt*(-1.0_rt - spzero)*dq(i,j,k,n);
+#if  PRIM_SPECIES_HAVE_SOURCES
+        qp(i,j,k,n) += 0.5_rt*dt*srcQ(i,j,k,n);
+#endif
+
```

Comment:
Looks like the closing bracket is misplaced here for the right state validity check.

Reply:
indeed.  Good catch.